agent/consul/state/catalog.go

```diff
@@ -365,21 +380,44 @@ func (s *Store) ensureNodeTxn(tx *memdb.Txn, idx uint64, node *structs.Node) err
 		if existing != nil {
 			n = existing.(*structs.Node)
 			if n.Node != node.Node {
-				return fmt.Errorf("node ID %q for node %q aliases existing node %q",
-					node.ID, node.Node, n.Node)
+				// Lets first get all nodes and check whether name do match
+				dupNameError := s.ensureNoNodeWithSimilarNameTxn(tx, node)
+				if dupNameError != nil {
+					return fmt.Errorf("Error while renaming Node ID: %q: %s", node.ID, dupNameError)
+				}
+				// We are actually renaming a node, remove its reference first
+				err := s.deleteNodeTxn(tx, idx, n.Node)
+				if err != nil {
+					return fmt.Errorf("Error while renaming Node ID: %q from %s to %s",
+						node.ID, n.Node, node.Node)
+				}
+			}
+		} else {
```

Comment:
The else statement here is one of the problematic ones. I would recommend replacing with this statement.

```
}
// TODO: else statement missing to ensure name is not already taken by another node
// If added right now it could break existing behavior which allows for adding
// a UUID to a node that doesn't have one when the name is the same.
```

Reply:
@mkeeler I think the else statement is now SAFE since `s.ensureNoNodeWithSimilarNameTxn` will not match any node without an ID. It means it will break ONLY if: a node having an ID clashes (case insensitive) with another node having an ID.

In all other cases, the behaviour is unchanged as shown in tests `deprecatedEnsureNodeWithoutIDCanRegister` of agent/consul/state/catalog_test.go.

# What works:

## When request has an ID:
* renaming a node when having an ID, without colliding with the name of another node (case insensitive)
* creating a node with an ID for which the name is not taken by anybody (case insensitive)

## When request has not ID (more permissive regarding case insensitivity)
* Create a new node without ID with name not colliding
* Remove a node ID of a node with an ID if name EXACTLY (case sensitive) matches
* Create a new node without ID with name colliding with case insensitive differences

# What does not work anymore due to this change:
* Create a node with an ID, but this name collides with another one (case insensitive) having an ID (works if other node DOES not have any ID)

# Previously, given the following configuration

id42 - node1

Registering same name with different ID is blocked
id43 - node2 => previously would 'steal' node2 to id42    ==> now blocked
id43 - NoDe2 => previously would create a second node (id42: node2 ; id43: NoDe2) => the big crash I had described here https://github.com/hashicorp/consul/issues/4413 => now the registration is blocked and id43 won't be able to join

# Conclusion

* A request with an ID can "steal" a node without any ID, but cannot steal a node with an ID if ID is different
* A request without any ID can do whatever it wants

So, basically, the ONLY new possible issue we have is:
 - if you change manually the ID of a node, this node won't be able to join the cluster
 - if 2 nodes have different IDs and case insensitive differences, only ONE can join, which is good new for DNS as well as it avoid the DB corruption I had.

What do you think of this ?

Reply:
I think I understand now.

We will error out if both if an existing node with a uuid != new uuid where there is a name match but we will still allow adding a uuid to a node that doesn't have one. You just moved the logic to only enforce it when they both have uuids into that function.